transport/tchannel/handler.go

```diff
@@ -188,12 +188,26 @@ func (h handler) callHandler(ctx context.Context, call inboundCall, responseWrit
 		ctx = tchannel.ExtractInboundSpan(ctx, tcall.InboundCall, headers.Items(), tracer)
 	}
 
+	buf := bufferpool.Get()
+	defer bufferpool.Put(buf)
+
 	body, err := call.Arg3Reader()
 	if err != nil {
 		return err
 	}
 	defer body.Close()
-	treq.Body = body
+
+	_, err = buf.ReadFrom(body)
+	if err != nil {
+		return err
+	}
+
+	err = body.Close()
+	if err != nil {
+		return err
+	}
+
+	treq.Body = buf
```

Comment:
I am not very convinced of doing this here. This extra steps will add some pressure on the GC and Memory, now for each request we are copying the entire io only to get its size. This seems a bit too much.

Did you explore the possibility of taking the size of the request and response in the `encoding` pk. IO are already being read there for response and request, maybe we can reuse this location to save the size of the request/response to be used later on by the middleware. 

We have the public method `func (h jsonHandler) Handle(ctx context.Context, treq *transport.Request, rw transport.ResponseWriter) error` in the encoding pkg.

Reply:
(discussed offline)
I agree, this caused double copy in thrift encoding and recent change avoids second read in the encoding handler.  
It's necessary read the body in the transport and before encoding handler as observability middleware comes right before encoding handlers, which enables observability handler to access body size directly instead of encoding layer modifying the transport request. Also, JSON encoding handler does not have access to the body as `json.decoder` consumes reader directly.